lib/rjgit.rb

```diff
@@ -341,21 +334,116 @@ def commit(message, author, parents = nil, ref = nil, force = false)
       end
-      
+
+    end
+
+    class ApplyPatchToIndex < RJGit::Plumbing::Index
+
+      import org.eclipse.jgit.patch.Patch
+      import org.eclipse.jgit.diff.DiffEntry
+
+      ADD    = org.eclipse.jgit.diff.DiffEntry::ChangeType::ADD
+      COPY   = org.eclipse.jgit.diff.DiffEntry::ChangeType::COPY
+      MODIFY = org.eclipse.jgit.diff.DiffEntry::ChangeType::MODIFY
+      DELETE = org.eclipse.jgit.diff.DiffEntry::ChangeType::DELETE
+      RENAME = org.eclipse.jgit.diff.DiffEntry::ChangeType::RENAME
+
+      # Take the result of RJGit::Porcelain.diff with options[:patch] = true and return a patch String
+      def self.diffs_to_patch(diffs)
+        result = ""
+        diffs.each {|diff| result << diff[:patch]}
+        result
+      end
+
+      def initialize(repository, patch, ref = Constants::HEAD)
+        super(repository)
+        @ref   = ref
+        @patch = Patch.new
+        @patch.parse(ByteArrayInputStream.new(patch.to_java_bytes))
+        raise_patch_apply_error unless @patch.getErrors.to_a.empty?
+        @current_tree = Commit.find_head(@jrepo, ref).tree
+      end
+
+      def commit(message, author, parents = nil, force = false)
+        super(message, author, parents, @ref, force)
+      end
+
+      def build_map
+        raise_patch_apply_error if @patch.getFiles.to_a.empty?
+        @patch.getFiles.each do |file_header|
+          case file_header.getChangeType
+          when ADD
+            add(file_header.getNewPath, apply('', file_header))
+          when MODIFY
+            add(file_header.getOldPath, apply(getData(file_header.getOldPath), file_header))
+          when DELETE
+            delete(file_header.getOldPath)
+          when RENAME
+            delete(file_header.getOldPath)
+            add(file_header.getNewPath, getData(file_header.getOldPath))
+          when COPY
+            add(file_header.getNewPath, getData(file_header.getOldPath))
+          end
+        end
+      end
+
+      private
+
+      def raise_patch_apply_error
+        raise Java::OrgEclipseJgitApiErrors::PatchApplyException.new('Patch failed to apply')
+      end
+
+      def getData(path)
+        begin
+          (@current_tree / path).data
+        rescue
+          raise_patch_apply_error
+        end
+      end
+
+      def hunk_sanity_check(hunk, hunk_line, pos, newLines)
+        raise_patch_apply_error unless newLines[hunk.getNewStartLine - 1 + pos] == hunk_line[1..-1]
+      end
+
+      def apply(original, file_header)
+        newLines = original.lines
+        file_header.getHunks.each do |hunk|
+          length = hunk.getEndOffset - hunk.getStartOffset
+          buffer_text = hunk.getBuffer.to_s.slice(hunk.getStartOffset, length)
+          pos = 0
+          buffer_text.each_line do |hunk_line|
+            case hunk_line[0]
+              when ' '
+                hunk_sanity_check(hunk, hunk_line, pos, newLines)
+                pos += 1
+              when '-'
```

Comment:
Why should `pos` not be incremented in this case?

Reply:
It's been a while since I've implemented this, but I think it's because the next hunk line will be at `pos`, and not `pos + 1` after the delete:

```
line 1, pos = 0
- line 2, pos = 1
line 3, pos = 2
```
After removing line 2, line 3's position is 1.

Reply:
Note this just follows JGit's implementation here: https://github.com/eclipse/jgit/blob/master/org.eclipse.jgit/src/org/eclipse/jgit/api/ApplyCommand.java